File: shuffle_cards.cpp
#include <iostream>
#include <stdlib.h>
#include <cstring>

/* This solution should not use STL */

void setDeck(int cards[], int numOfCards)
{
	/* Function to set the deck of cards initially */
	for (int i=0;i<numOfCards; i++)
	{
		cards[i] = i+1;
	}
}

void displayDeck(int cards[], int numOfCards)
{
	/* Function to display the deck of cards */
	for (int i =0; i < numOfCards; i++)
		std::cout << cards[i] <<" ";
	std::cout << std::endl;
} 

bool checkDeck(int cards[], int numOfCards)
{
	/* Function to check if the deck is in order */
	for (int i=0; i<numOfCards;i++)
	{
		if  (cards[i] != i+1)
		{
			return false;
		}
	}
	return true;
}

int*  ShuffleDeck(int cards[], int numOfCards)
{

	/*Function executes one "round" of shuffle & 
	 * Returns the Shuffled Deck
	 */
	int* result = new int[numOfCards];
	int size = numOfCards;

	for (int j=0; j < numOfCards; j++)
	{
		// Putting the top cards into the result deck
		result[j] = cards[0];

		// Delete the top card from the list
		for (int k=0; k < numOfCards; k++)
		{
			cards[k] = cards[k+1];
		}

		// "Putting the next top of deck in the end"
		cards[size-1] = cards[0];

		// Delete the top card from the list
		for (int k=0; k < numOfCards; k++)
		{
			cards[k] = cards[k+1];
		}

		size -= 1;
	}
	return result;
}

int countRounds(int cards[], int numOfCards)
{
	bool inOrder = false;
	bool check = false;
	int count = 0;

	int* temp = new int[numOfCards];
	while (not inOrder)
	{
		temp = ShuffleDeck(cards, numOfCards);
		count += 1;

		std::cout <<"The shuffled deck of cards are: ";
		displayDeck(temp, numOfCards);

		check = checkDeck(temp, numOfCards);
		if (check)
		{
			inOrder = true;
			return count;
		}

		else
		{
			cards = temp;
		}

	}
}

int main(int argc, char *argv[])
{
	int numOfCards;
	int numOfRounds;
	
	if (argc > 2)
	{
		std::cerr << "Usage: " << argv[0] << std::endl;
		return 1;
	}

	numOfCards = atoi(argv[1]);
	int cards[numOfCards];

	// Setting the Deck of Cards and Displaying the original deck
	setDeck(cards, numOfCards);
	std::cout <<"The orignal deck of cards are:  ";
	displayDeck(cards,numOfCards);

	//Shuffling the cards and checking for order
	numOfRounds = countRounds(cards, numOfCards);
	std::cout << "The number of rounds is: " << numOfRounds << std::endl;
	return 0;
}

File: shuffle_cards.py
"""
Solution Module to determine how many rounds it takes to restore deck to original order
How is shuffling done?
"""


def check_deck(deck):
    """
    @brief: Function to check if deck is in order
    :param: deck (type: list)
    """
    for i in range(len(deck)):
        if deck[i] != i+1:
            return False
    return True

def shuffle_deck(deck):
    """
    @brief: Function to shuffle the deck and return it after a single round
    :param: deck (type: list)
    """
    stack = []
    while len(deck) > 0:
        if len(deck) > 2:
            for i in range(0, len(deck), 2):
                stack.append(deck[i])
            del deck[::2]

        else:
            stack += deck[::-1]
            deck = []
    return stack

def get_number_of_rounds(num_of_cards):
    """
    @brief: Core Function get number of rounds it takes to restore original order
    @param: num_of_cards - number of cards, type: int
    """
    deck = [0] * num_of_cards
    for i in range(0, num_of_cards):
        deck[i] = i+1
    print("The deck of cards is:", deck)

    in_order = False
    count = 0

    while not in_order:
        result = shuffle_deck(deck)
        count += 1
        print("The shuffled deck is:", result)
        if check_deck(result):
            in_order = True
            return count
        else:
            deck = result

if __name__ == '__main__':
    for i in range(2, 11):
        num_of_rounds = get_number_of_rounds(i)
        print("The number of rounds is", num_of_rounds)
        print("****************************************")
